executor/src/witgen/affine_expression.rs

```diff
@@ -341,7 +431,10 @@ where
     T: FieldElement,
 {
     fn eq(&self, other: &Self) -> bool {
-        self.offset == other.offset && self.nonzero_coefficients().eq(other.nonzero_coefficients())
+        self.offset == other.offset
+            && self
+                .nonzero_coefficients()
```

Comment:
This relies on `nonzero_coefficients()` returning a sorted array, which is not the case if the expression is clean. The following test fails:

```rust
    #[test]
    pub fn test_equality() {
        let a = AffineExpression::<_, GoldilocksField> {
            coefficients: convert(vec![0, 1]),
            offset: 3.into(),
            clean: false,
        }
        .clean();
        let b = AffineExpression {
            coefficients: convert(vec![1, 0]),
            offset: 13.into(),
            clean: false,
        }
        .clean();
        assert_eq!(a.clone() + b.clone(), b.clone() + a.clone());
    }
```

Reply:
Ah nice, good catch!

Reply:
I first thought about inserting single elements into a sorted list, but it turns out that PartialEq is not really used in the code. But maybe it still makes sense to sort?